### PR TITLE
Fix code coverage.

### LINKED ISSUE #5265: grpc_coverage is not generating reports
@nicolasnoble @jtattermusch any ideas?

### PR DESCRIPTION
**Pull request by adelez**
Changed the code based on the print message below: 
"Multi-language **and** multi-config testing is not supported".
Hope it fixes #5265.

### CODE REVIEW

**Comment by adelez** [comment]
@ctiller

**Comment by adelez** [comment]
A few lines above:
if len(build_configs) > 1:
800	  for language in languages:
801	    if not language.supports_multi_config():
802	      print language, 'does not support multiple build configurations'
803	      sys.exit(1)

It looks like some languages do support multi config. I wonder if this condition should be removed completely.

File: tools/run_tests/run_tests.py [hunk 1]
```diff
@@ -810,7 +810,7 @@ def git(cmd, cwd=cwd):
   else:
     language_make_options = next(iter(languages)).make_options()
 
-if len(languages) != 1 or len(build_configs) != 1:
+if len(languages) != 1 and len(build_configs) != 1:
```
Comment (jtattermusch) [inline tools/run_tests/run_tests.py]: well, that's tough. This condition is here for a reason. I've split up testing docker images per-language a while ago and currently, run_tests.py can only use one docker image in a single run (and yeah, I forgot about code coverage test that requires a single umbrella image when doing the change - my bad).

This check is here to protect us against cases where we want to run more languages at a time (-l all), but that would require using more docker images which we currently don't support.

I'd be actually surprised if the coverage run would actually work with this "fix".

**Comment by adelez** [comment]
I tried locally and did get the reports.

**Comment by jtattermusch** [comment]
Something has to be wrong with that report.
If you look a few lines lower, the dockerfile used for the test run will be the one specified for the first language in the list:
https://github.com/adelez/grpc/blob/fix_code_coverage/tools/run_tests/run_tests.py#L831

so let's say it will pick up the cxx_jessie_x64 docker image (because it's going to be the first one in the list).  Then it will try to run tests for all the other languages  (node, ruby, csharp, ...) inside that docker image, but the node, ruby, csharp dependencies aren't even installed in that dockerimage - so there's no way this could possibly work. So the suite is either failing to indicate there's a problem or things are just weird.

What cmdline did you use when you tested it locally?

**Comment by adelez** [comment]
tools/run_tests/run_tests.py --use_docker --measure_cpu_costs -t -l all -c gcov -x report.xml -j 2

**Comment by adelez** [comment]
Actually I take the statement back the result in my local drive is an old one. I'll take a further look.

**Comment by adelez** [comment]
The command from the console output is 
tools/run_tests/run_tests.py --use_docker --measure_cpu_costs -t -l all -c gcov -x report.xml -j 2

Will building an umbrella image solve this problem?

**Comment by adelez** [comment]
Separate from #5265, if -l all is no longer supported by run_tests.py, 
argp.add_argument('-l', '--language',
                  choices=['all'] + sorted(_LANGUAGES.keys()),
                  nargs='+',
                  default=['all'])
should be modified to reflect this. If there's no objection, I'll go ahead make the change.

**Comment by jtattermusch** [comment]
we used to have an umbrella image "grpc_tests_multilang_x64", but we didn't like it much.
Here's my pull request https://github.com/grpc/grpc/pull/5121 that refactored our docker images.

I'm still not sure what needs to be done here. Reintroducing the multilang image and making sure gcov with -l all uses it would probably be a hotfix, but it doesn't really fit in the bigger picture, so I am not too supportive of this. More likely, we need to take a fresh look at things.

**Comment by jtattermusch** [comment]
@adelez  after https://github.com/grpc/grpc/pull/5369 gets merged, it should be possible to make tweaks to run_tests.py and use the multilang image (needs to be reintroduced and properly generated using templates) to run the  -l all -c gcov.

**Comment by jtattermusch** [comment]
I'll close this PR for now.

**Comment by adelez** [comment]
Thanks. I'll do.

**Pull request closed without merging**